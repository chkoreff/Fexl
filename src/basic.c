#include <value.h>

#include <basic.h>

value QI;
value QT;
value QF;
value QY;
value Qvoid;
value Qnull;
value Qonce;
value Qyield;

// (I x) = x
value type_I(value fun, value f)
	{
	(void)fun;
	(void)f;
	return hold(f->R);
	}

// (T x y) = x
value type_T(value fun, value f)
	{
	if (fun->L == 0) return keep(fun,f);
	return hold(fun->R);
	}

// (F x y) = y
value type_F(value fun, value f)
	{
	if (fun->L == 0) return keep(fun,f);
	return hold(f->R);
	}

// (Y x) = (x (Y x))
value type_Y(value fun, value f)
	{
	return A(hold(f->R),A(hold(fun),hold(f->R)));
	}

// (void x) = void
value type_void(value fun, value f)
	{
	(void)fun;
	(void)f;
	return hold(Qvoid);
	}

// ((pair x y) h) = (h x y)
value type_pair(value fun, value f)
	{
	value x = hold(fun->L);
	value y = hold(fun->R);
	value h = hold(f->R);
	return A(A(h,x),y);
	}

// ((list x y) h) = (pair x y)
value type_list(value fun, value f)
	{
	value x = hold(fun->L);
	value y = hold(fun->R);
	(void)f;
	return V(type_pair,x,y);
	}

// (cons x y) = (list x y)
value type_cons(value fun, value f)
	{
	if (fun->L == 0) return keep(fun,f);
	{
	value x = hold(fun->R);
	value y = hold(f->R);
	return V(type_list,x,y);
	}
	}

// (null x y) = x
value type_null(value fun, value f)
	{
	return V(type_T,hold(fun),hold(f->R));
	}

// (eval x f) = (f y), where y is the final value of x.
value type_eval(value fun, value f)
	{
	if (fun->L == 0) return keep(fun,f);
	return A(hold(f->R),arg(fun->R));
	}

// Evaluate x once, replacing the right side with the final value.
value type_once(value fun, value f)
	{
	(void)fun;
	return (f->R = hold(eval(f->R)));
	}

// (yield x f) = (f x)  Used for returning an unevaluated function.
value type_yield(value fun, value f)
	{
	if (fun->L == 0) return keep(fun,f);
	return A(hold(f->R),hold(fun->R));
	}

value yield(value x)
	{
	return V(type_yield,hold(Qyield),x);
	}

// Return either no or (yes x).
value maybe(value x)
	{
	if (x == 0)
		return hold(QT);
	else
		return V(type_T,hold(QT),yield(hold(x)));
	}

value boolean(int x)
	{
	return hold(x ? QT : QF);
	}

value pair(value x, value y)
	{
	return V(type_pair,x,y);
	}

value type_is_defined(value fun, value f)
	{
	(void)fun;
	return boolean(f->R->T != type_void);
	}

value type_is_undef(value fun, value f)
	{
	(void)fun;
	return boolean(f->R->T == type_void);
	}

value op_is_type(value fun, value f, type t)
	{
	value x = arg(f->R);
	f = boolean(x->T == t);
	drop(x);
	(void)fun;
	return f;
	}

value type_is_void(value fun, value f)
	{
	return op_is_type(fun,f,type_void);
	}

static value op_predicate(value fun, value f, int op(value))
	{
	value x = arg(f->R);
	f = boolean(op(x));
	drop(x);
	(void)fun;
	return f;
	}

static int is_good(value x)
	{
	return (x->T != type_void);
	}

static int is_bool(value x)
	{
	return (x->T == type_T || x->T == type_F) && !x->L;
	}

static int is_list(value x)
	{
	return x->T == type_list || x->T == type_null;
	}

value type_is_good(value fun, value f) { return op_predicate(fun,f,is_good); }
value type_is_bool(value fun, value f) { return op_predicate(fun,f,is_bool); }
value type_is_list(value fun, value f) { return op_predicate(fun,f,is_list); }

// \chain=(\\a\\b \x \v=(a x) is_defined v v (b x))
// Return (a x) if that is defined, otherwise return (b x).
value type_chain(value fun, value f)
	{
	if (fun->L == 0) return keep(fun,f);
	if (fun->L->L == 0) return keep(fun,f);
	{
	value x = arg(f->R);
	value a = (fun->L->R = eval(fun->L->R));
	value v = eval(A(hold(a),x));

	if (v->T != type_void)
		return v;

	drop(v);
	{
	value b = (fun->R = eval(fun->R));
	return A(hold(b),hold(x));
	}
	}
	}

// Create a nullary function, supplying a dummy argument.
value Q0(type T)
	{
	return A(Q(T),hold(QI));
	}

void beg_basic(void)
	{
	QI = Q(type_I);
	QT = Q(type_T);
	QF = Q(type_F);
	QY = Q(type_Y);
	Qvoid = Q(type_void);
	Qnull = Q(type_null);
	Qonce = Q(type_once);
	Qyield = Q(type_yield);
	}

void end_basic(void)
	{
	drop(QI);
	drop(QT);
	drop(QF);
	drop(QY);
	drop(Qvoid);
	drop(Qnull);
	drop(Qonce);
	drop(Qyield);
	}
