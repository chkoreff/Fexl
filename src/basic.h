extern struct value *type_C(struct value *);
extern struct value *type_S(struct value *);
extern struct value *type_I(struct value *);
extern struct value *type_R(struct value *);
extern struct value *type_L(struct value *);
extern struct value *type_Y(struct value *);
extern struct value *type_F(struct value *f);
extern struct value *type_query(struct value *);
