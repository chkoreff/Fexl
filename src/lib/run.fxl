#use_new_syntax
# Run a child function with the given input.  Collect the stdout as it runs,
# and collect the stderr after it terminates.  Return the output, error, and
# status.
\run_function=
	(\child_fn\input
	spawn child_fn \child_in\child_out\child_err

	fput child_in input
	fclose child_in

	\\out=(file_content child_out)

	\status
	\\err=(file_content child_err)
	{ out err status }
	)

# Run the function (exec argv).
\run_program=
	(\argv
	run_function (exec argv)
	)

# Run a function with the given input, returning the output as a string.  If
# there was any error, show that on stderr and die.
\run_function_filter=
	(\child_fn\input
	run_function child_fn input \out\err\status
	if (ne status 0) (error err)
	out
	)

# Run (exec argv) as a filter.
\run_filter=
	(\argv
	run_function_filter (exec argv)
	)

# Grab the output of a function, passing in null input.
\grab_output=
	(\child_fn
	run_function_filter child_fn ""
	)

# LATER 20230425 Use a pid file instead, because if you're running the browser
# on the same machine as the server, this can kill the browser too.
\stop_server=
	(\ip\port
	\argv=["/usr/bin/lsof" (to_str ["-i4TCP@"ip":"port]) "-t"]
	\\out=(run_function (exec argv) "" \out\err\status out)
	\\list_pid=(map str_num; filter (ne ""); split NL out)

	each list_pid
		(\pid
		kill pid 2 # SIGINT
		)
	)

\run_server=
	(\option\ip\port\error_log\interact

	\start_server=(start_server ip port error_log interact)
	\stop_server=(stop_server ip port)

	\ok=
		(
		is_void option F;
		eq option "start" T;
		eq option "stop" T;
		F
		)

	if (not ok) (error ~ Please specify either "start" or "stop".~)

	stop_server
	if (eq option "start") start_server
	)

\form
def "run_function" run_function;
def "run_program" run_program;
def "run_function_filter" run_function_filter;
def "run_filter" run_filter;
def "grab_output" grab_output;
def "stop_server" stop_server;
def "run_server" run_server;
form
