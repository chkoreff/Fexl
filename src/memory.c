#include <stdio.h>

#include <die.h>
#include <file.h>
#include <memory.h>
#include <stdlib.h> /* malloc free */

/* Track the amount of memory used so I can detect memory leaks.  Normally this
is impossible but I check it anyway in case of software error. */
static unsigned long cur_blocks = 0;
unsigned long cur_bytes = 0;

/* Return a new span of memory of size num_bytes, or die if not possible. */
void *new_memory(unsigned long num_bytes)
	{
	if (num_bytes == 0) die("NEW0");
	{
	void *data = malloc(num_bytes);
	if (data == 0)
		die("The program ran out of memory.");
	cur_blocks++;
	cur_bytes += num_bytes;
	return data;
	}
	}

/* Free a previously allocated span of memory. */
void free_memory(void *data, unsigned long num_bytes)
	{
	if (!data) die("NFREE");
	if (cur_blocks == 0 || cur_bytes < num_bytes)
		die("XFREE");

	free(data);
	cur_blocks--;
	cur_bytes -= num_bytes;
	}

/* Detect any final memory leak, which should never happen. */
void end_memory(void)
	{
	if (cur_blocks || cur_bytes)
		{
		fput(stderr,"LEAK");
		fput(stderr," ");fput_ulong(stderr,cur_blocks);
		fput(stderr," ");fput_ulong(stderr,cur_bytes);
		die("");
		}
	}
