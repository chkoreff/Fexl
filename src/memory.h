extern unsigned long cur_bytes;
extern void *new_memory(unsigned long num_bytes);
extern void free_memory(void *data, unsigned long num_bytes);
extern void end_memory(void);
