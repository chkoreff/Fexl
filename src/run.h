extern int main_argc;
extern char **main_argv;
extern long argi;
extern int main_envc();
extern char **main_envp;
extern char *error;
extern long line;
extern long max_depth;
extern long cur_depth;
extern long max_steps;
extern long cur_steps;
extern void beg_run(int argc, char *argv[], char *envp[]);
