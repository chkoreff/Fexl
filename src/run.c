#include "run.h"

/* Command line arguments */
int main_argc;
char **main_argv;

static int cache_envc = -1; /* auto-initialized count of envp entries */
char **main_envp;

void beg_run(int argc, char *argv[], char *envp[])
	{
	main_argc = argc;
	main_argv = argv;
	main_envp = envp;
	}

int main_envc(void)
	{
	if (cache_envc < 0)
		{
		cache_envc = 0;
		if (main_envp != 0)
			while (1)
				{
				char *s = main_envp[cache_envc];
				if (s == 0) break;
				cache_envc++;
				}
		}

	return cache_envc;
	}
