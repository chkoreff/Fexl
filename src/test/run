#!/bin/sh
./build || exit
. ./safe

run_fexl()
	{
	local file="$1"
	echo -n "===="
	if [ "$file" != "" ]; then echo -n " $file"; fi
	echo
	../bin/fexl $file 2>&1
	echo "exit $?"
	}

run_fexl test/a1.fxl

(
cat <<"EOF"
abc
defg

hij
åabcüdef
üä 1≠0 包子
EOF
) | run_fexl test/a2.fxl

# Test stdout versus stderr.
echo "== stdout:"
../bin/fexl test/a3.fxl 2>&1 2>/dev/null
echo "== stderr:"
../bin/fexl test/a3.fxl 2>&1 1>/dev/null

run_fexl test/a4.fxl
run_fexl test/a5.fxl
run_fexl test/a6.fxl
run_fexl test/a7.fxl
run_fexl test/a8.fxl
run_fexl test/a9.fxl
run_fexl test/b10.fxl
run_fexl test/b11.fxl
run_fexl test/b12.fxl
run_fexl test/b13.fxl
run_fexl test/b14.fxl
run_fexl test/b15.fxl
run_fexl test/b16.fxl
run_fexl test/b17.fxl
run_fexl test/b18.fxl
run_fexl test/b19.fxl
run_fexl test/b20.fxl
run_fexl test/chars.fxl
run_fexl test/date.fxl
run_fexl test/sat.fxl
run_fexl test/crypto.fxl

(
cat <<"EOF"
abcABC
jjja小
åüä≠
包子
EOF
) | run_fexl test/get_byte.fxl

# Test the "die" function.
(
cat <<"EOF"
say "hi"
die
say "bye"
EOF
) | run_fexl

# Test some syntax errors.

echo -n "\"" | run_fexl

echo -n '
"a
b' | run_fexl

echo -n "~| a" | run_fexl

echo -n "

~END
abc~EN" | run_fexl

echo -n "~" | run_fexl

echo -n "~
" | run_fexl

echo -n "~EN" | run_fexl

echo -n " = y" | run_fexl

echo -n "ab cd e\\" | run_fexl

echo -n "\\x=" | run_fexl

echo -n "(
a
(b" | run_fexl

echo -n "(
a
(b)" | run_fexl

echo -n "[
a
[b" | run_fexl

echo -n "{
a
{b" | run_fexl

echo -n "(ab cd e)
)" | run_fexl

(
cat <<"EOF"
say (+ 4 6)
( a b c ) ( b c d )
( a b c ) ( b c d )
put (- 12 7.2) nl
x
EOF
) | run_fexl

(
cat <<"EOF"
evaluate (define "b" (say "b"); void) \; a b
a c
c
EOF
) | run_fexl

run_fexl missing.fxl

(
cat <<"EOF"
parse_file "missing.fxl"
EOF

) | run_fexl

# Here I test setting ulimit to terminate misbehaving programs promptly.
test_misbehavior=false
# To enable the test, unpound the next line.
#test_misbehavior=true

if $test_misbehavior; then

(
ulimit -St 1
echo -n "@ I" | run_fexl
# CPU time limit exceeded
)

(
ulimit -Ss 1000
echo -n "@ @" | run_fexl
# Stack overflow (Segmentation fault)
)

(
ulimit -Sv 100000
echo -n "\S=(\x\y\z x z; y z) @ S S S" | run_fexl
# The program ran out of memory.
)

fi

exit 0
