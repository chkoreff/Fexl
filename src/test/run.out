beg f = ((C S) C)
end f = ((C S) C)
end g = S

beg f = C
end f = C
end g = C

beg f = (C C)
end f = (C C)
end g = (C C)

beg f = ((C C) C)
end f = ((C C) C)
end g = C

beg f = (((C C) C) ((C C) C))
end f = (C ((C C) C))
end g = (C ((C C) C))

beg f = (((S C) S) (C C))
end f = (((S C) S) (J (C C)))
end g = (C C)

beg f = ((C ((C C) C)) C)
end f = ((C ((C C) C)) C)
end g = C

beg f = (((S C) S) (((S C) S) (((S C) S) (((S C) S) ((C (S C)) C)))))
end f = (((S C) S) (J (S C)))
end g = (S C)

beg f = ((((C C) (C C)) ((C C) (C C))) (((C C) (C C)) ((C C) (C C))))
end f = ((C ((C C) (C C))) (C ((C C) (C C))))
end g = C

beg f = (((((S S) S) S) C) C)
end f = (((S (J C)) ((S (J S)) (J C))) (J C))
end g = C

beg f = (((S I) I) C)
end f = (((S I) I) (J C))
end g = (C (I (J C)))

beg f = (((S ((S C) C)) ((C C) C)) ((C S) C))
end f = (((S ((S C) C)) ((C C) C)) (J S))
end g = (S (((C C) C) (J S)))

beg f = (say "ping!")
end f = (say "ping!")
end g = (say "ping!")

beg f = (I (say "ping!"))
end f = (I (say "ping!"))
end g = (say "ping!")

beg f = C
end f = C
end g = C

beg f = (C C)
end f = (C C)
end g = (C C)

beg f = ((C C) C)
end f = ((C C) C)
end g = C

beg f = ((C S) C)
end f = ((C S) C)
end g = S

beg f = (((R ((R ((R ((L I) I)) ((S R) I))) ((S R) I))) ((S R) ((S R) I))) (say "ping!"))
ping!
ping!
ping!
ping!
ping!
ping!
ping!
ping!
ping!
ping!
ping!
ping!
end f = (((R ((R ((R ((L I) I)) ((S R) I))) ((S R) I))) ((S R) ((S R) I))) (say "ping!"))
end g = I

beg f = @
end f = @
end g = @

beg f = (@ C)
end f = (@ C)
end g = (C (@ C))

beg f = (@ I)
end f = (@ I)
end g = (I (@ I))

beg f = (@ @)
end f = (@ @)
end g = ((@ (@ @)) (@ (@ @)))

beg f = (((@ S) S) S)
end f = (((S (@ S)) S) (J S))
end g = (((S (@ S)) S) (J S))

beg f = "abc"
end f = "abc"
end g = "abc"

beg f = (I "abc")
end f = (I "abc")
end g = "abc"

beg f = ("abc" "de")
end f = ("abc" "de")
end g = ("abc" "de")

beg f = ((concat "abc") "de")
end f = ((concat "abc") "de")
end g = "abcde"

beg f = ((I (I "abc")) "de")
end f = ("abc" "de")
end g = ("abc" "de")

beg f = ((concat (I "abc")) "de")
end f = ((concat "abc") "de")
end g = "abcde"

beg f = ((concat (C "abc")) "de")
end f = ((concat (C "abc")) "de")
end g = ((concat (C "abc")) "de")

beg f = ((concat ((I I) "abc")) ((concat "") "de"))
end f = ((concat "abc") "de")
end g = "abcde"

beg f = 3.14159265358979
end f = 3.14159265358979
end g = 3.14159265358979

beg f = 42
end f = 42
end g = 42

beg f = (length "12345")
end f = (length "12345")
end g = 5

beg f = (length ((concat (I "abc")) ((concat "") "d")))
end f = (length "abcd")
end g = 4

beg f = (((slice ((C ((concat "abcd") "efghi")) S)) 2) 5)
end f = (((slice "abcdefghi") 2) 5)
end g = "cdefg"

beg f = ((+ -14.87) 482.73)
end f = ((+ -14.87) 482.73)
end g = 467.86

beg f = ((/ ((+ 1) (sqrt 5))) 2)
end f = ((/ 3.23606797749979) 2)
end g = 1.61803398874989

beg f = ((C "yes") "no")
end f = ((C "yes") "no")
end g = "yes"

beg f = ((F "yes") "no")
end f = (I "no")
end g = "no"

beg f = ((ge (I 5.9)) 5.8)
end f = ((ge 5.9) 5.8)
end g = C

beg f = ((lt 5.9) (I 5.8))
end f = ((lt 5.9) 5.8)
end g = F

beg f = ((ge (I "ab")) "a")
end f = ((ge "ab") "a")
end g = C

beg f = ((lt (I "ab")) (I "a"))
end f = ((lt "ab") "a")
end g = F

beg f = ((eq (I "x")) "x")
end f = ((eq "x") "x")
end g = C

beg f = ((eq (I -7.2)) -7.2)
end f = ((eq -7.2) -7.2)
end g = C

beg f = ((lt (I 5.9)) (I "ab"))
end f = ((lt 5.9) "ab")
end g = ((lt 5.9) "ab")

beg f = ((lt (I "ab")) (I 5.9))
end f = ((lt "ab") 5.9)
end g = ((lt "ab") 5.9)

beg f = ((eq (I S)) (I S))
end f = ((eq S) (I S))
end g = ((eq S) (I S))

beg f = (num_str ((/ ((+ 1) (sqrt 5))) 2))
end f = (num_str 1.61803398874989)
end g = "1.61803398874989"

beg f = (str_num ((concat "+") ((concat "1.") "61803398874989")))
end f = (str_num "+1.61803398874989")
end g = 1.61803398874989

beg f = (str_num (I 5.8))
end f = (str_num 5.8)
end g = (str_num 5.8)

beg f = (str_num (I S))
end f = (str_num S)
end g = (str_num S)

beg f = (((S ((S "x0") "x1")) "y") "z")
end f = (((S ((S "x0") "x1")) "y") (J "z"))
end g = ((("x0" (J "z")) ("x1" (J "z"))) ("y" (J "z")))

beg f = {x}
end f = {x}
end g = {x}

beg f = {"x"}
end f = {"x"}
end g = {"x"}

beg f = ({x} {"x"})
end f = {{x} {"x"}}
end g = {{x} {"x"}}

beg f = ({"x"} {x})
end f = {{"x"} {x}}
end g = {{"x"} {x}}

beg f = (I {{(I I) {x}} {C {y}}})
end f = (I {{(I I) {x}} {C {y}}})
end g = {{(I I) {x}} {C {y}}}

beg f = I
end f = I
end g = I

beg f = {C {x}}
end f = {C {x}}
end g = {C {x}}

beg f = {(L I) {y}}
end f = {(L I) {y}}
end g = {(L I) {y}}

beg f = {x}
end f = {x}
end g = {x}

beg f = {C {{y} {y}}}
end f = {C {{y} {y}}}
end g = {C {{y} {y}}}

beg f = {{R {y}} ((S I) I)}
end f = {{R {y}} ((S I) I)}
end g = {{R {y}} ((S I) I)}

beg f = {(L ((S I) I)) {y}}
end f = {(L ((S I) I)) {y}}
end g = {(L ((S I) I)) {y}}

beg f = {{S {(S I) {(L I) {"abc"}}}} {{S {y}} I}}
end f = {{S {(S I) {(L I) {"abc"}}}} {{S {y}} I}}
end g = {{S {(S I) {(L I) {"abc"}}}} {{S {y}} I}}

beg f = ?
end f = ?
end g = ?

beg f = ?
end f = ?
end g = ?

beg f = ?
end f = ?
end g = ?

beg f = put
end f = put
end g = put

beg f = (put (I "XYZ"))
end f = (put (I "XYZ"))
end g = (put (I "XYZ"))

beg f = ((put (I "XYZ")) {next})
XYZend f = ((put "XYZ") {next})
end g = {next}

beg f = ((put (I 3.14159265358979)) {next})
3.14159265358979end f = ((put 3.14159265358979) {next})
end g = {next}

beg f = ((put (C I)) {next})
end f = ((put (C I)) {next})
end g = ((put (C I)) {next})

beg f = nl
end f = nl
end g = nl

beg f = (nl {next})

end f = (nl {next})
end g = {next}

beg f = say
end f = say
end g = say

beg f = (say (I "XYZ"))
end f = (say (I "XYZ"))
end g = (say (I "XYZ"))

beg f = ((say (I "XYZ")) {next})
XYZ
end f = ((say "XYZ") {next})
end g = {next}

beg f = ((say (I 3.14159265358979)) {next})
3.14159265358979
end f = ((say 3.14159265358979) {next})
end g = {next}

beg f = ((say (C I)) {next})
end f = ((say (C I)) {next})
end g = ((say (C I)) {next})

beg f = (parse_string "")
end f = (parse_string "")
end g = ((R C) ((L I) (C ((L I) I))))

beg f = (parse_string " 	
x y")
end f = (parse_string " 	
x y")
end g = ((R C) ((L I) ((S ((R R) ((L ((L ((L I) F)) "y")) 2))) ((S ((R R) ((L ((L ((L I) F)) "x")) 2))) (C ((L I) I))))))

beg f = (parse_string "ab cd e")
end f = (parse_string "ab cd e")
end g = ((R C) ((L I) ((S ((R R) ((L ((L ((L I) F)) "e")) 1))) ((S ((R R) ((L ((L ((L I) F)) "cd")) 1))) ((S ((R R) ((L ((L ((L I) F)) "ab")) 1))) (C ((L I) I)))))))

beg f = (parse_string "abcdefghi;")
end f = (parse_string "abcdefghi;")
end g = ((R C) ((L I) ((S ((R R) ((L ((L ((L I) F)) "abcdefghi")) 1))) (C ((L I) ((L I) I))))))

beg f = (parse_string """")
end f = (parse_string """")
end g = ((R C) ((L I) ((S ((R R) ((L ((L ((L I) C)) "")) 1))) (C ((L I) I)))))

beg f = (parse_string ""a"")
end f = (parse_string ""a"")
end g = ((R C) ((L I) ((S ((R R) ((L ((L ((L I) C)) "a")) 1))) (C ((L I) I)))))

beg f = (parse_string ""abc
defg"")
end f = (parse_string ""abc
defg"")
end g = ((R C) ((L I) ((S ((R R) ((L ((L ((L I) C)) "abc
defg")) 1))) (C ((L I) I)))))

beg f = (parse_string """)
end f = (parse_string """)
end g = (C ((L ((L I) "Unclosed string")) 1))

beg f = (parse_string "
"a
b
")
end f = (parse_string "
"a
b
")
end g = (C ((L ((L I) "Unclosed string")) 2))

beg f = (parse_string "~| |")
end f = (parse_string "~| |")
end g = ((R C) ((L I) ((S ((R R) ((L ((L ((L I) C)) "")) 1))) (C ((L I) I)))))

beg f = (parse_string "~| a|")
end f = (parse_string "~| a|")
end g = ((R C) ((L I) ((S ((R R) ((L ((L ((L I) C)) "a")) 1))) (C ((L I) I)))))

beg f = (parse_string "~| a")
end f = (parse_string "~| a")
end g = (C ((L ((L I) "Unclosed string")) 1))

beg f = (parse_string "~END aEND")
end f = (parse_string "~END aEND")
end g = ((R C) ((L I) ((S ((R R) ((L ((L ((L I) C)) "a")) 1))) (C ((L I) I)))))

beg f = (parse_string "

~END
abcEN")
end f = (parse_string "

~END
abcEN")
end g = (C ((L ((L I) "Unclosed string")) 3))

beg f = (parse_string "~")
end f = (parse_string "~")
end g = (C ((L ((L I) "Incomplete string terminator")) 1))

beg f = (parse_string "~EN")
end f = (parse_string "~EN")
end g = (C ((L ((L I) "Incomplete string terminator")) 1))

beg f = (parse_string "~ABC abcdABAABC")
end f = (parse_string "~ABC abcdABAABC")
end g = ((R C) ((L I) ((S ((R R) ((L ((L ((L I) C)) "abcdABA")) 1))) (C ((L I) I)))))

beg f = (parse_string "~ABCD AABABCABCD")
end f = (parse_string "~ABCD AABABCABCD")
end g = ((R C) ((L I) ((S ((R R) ((L ((L ((L I) C)) "AABABC")) 1))) (C ((L I) I)))))

beg f = (parse_string " = y")
end f = (parse_string " = y")
end g = (C ((L ((L I) "Missing symbol declaration before '='")) 1))

beg f = (parse_string "ab cd e;")
end f = (parse_string "ab cd e;")
end g = ((R C) ((L I) ((S ((R R) ((L ((L ((L I) F)) "e")) 1))) ((S ((R R) ((L ((L ((L I) F)) "cd")) 1))) ((S ((R R) ((L ((L ((L I) F)) "ab")) 1))) (C ((L I) ((L ((R L) (R L))) I))))))))

beg f = (parse_string "ab cd e;f g")
end f = (parse_string "ab cd e;f g")
end g = ((R C) ((L I) ((S ((R R) ((L ((L ((L I) F)) "g")) 1))) ((S ((R R) ((L ((L ((L I) F)) "f")) 1))) ((S ((R R) ((L ((L ((L I) F)) "e")) 1))) ((S ((R R) ((L ((L ((L I) F)) "cd")) 1))) ((S ((R R) ((L ((L ((L I) F)) "ab")) 1))) (C ((L I) (R (R R)))))))))))

beg f = (parse_string "ab cd e;f g ; hi j;")
end f = (parse_string "ab cd e;f g ; hi j;")
end g = ((R C) ((L I) ((S ((R R) ((L ((L ((L I) F)) "j")) 1))) ((S ((R R) ((L ((L ((L I) F)) "hi")) 1))) ((S ((R R) ((L ((L ((L I) F)) "g")) 1))) ((S ((R R) ((L ((L ((L I) F)) "f")) 1))) ((S ((R R) ((L ((L ((L I) F)) "e")) 1))) ((S ((R R) ((L ((L ((L I) F)) "cd")) 1))) ((S ((R R) ((L ((L ((L I) F)) "ab")) 1))) (C ((L I) ((L ((R L) ((R (R L)) ((R (R (R R))) ((R (R (R R))) ((R (R (R R))) (R (R R)))))))) ((L (_ _)) _)))))))))))))

beg f = (parse_string "ab cd e\\fg h")
end f = (parse_string "ab cd e\\fg h")
end g = ((R C) ((L I) ((S ((R R) ((L ((L ((L I) F)) "e")) 1))) ((S ((R R) ((L ((L ((L I) F)) "cd")) 1))) ((S ((R R) ((L ((L ((L I) F)) "ab")) 1))) (C ((L I) I)))))))

beg f = (parse_string "ab cd e\x")
end f = (parse_string "ab cd e\x")
end g = ((R C) ((L I) ((S ((R R) ((L ((L ((L I) F)) "e")) 1))) ((S ((R R) ((L ((L ((L I) F)) "cd")) 1))) ((S ((R R) ((L ((L ((L I) F)) "ab")) 1))) (C ((L I) ((L ((R L) (R L))) (C I)))))))))

beg f = (parse_string "ab cd e\")
end f = (parse_string "ab cd e\")
end g = (C ((L ((L I) "Missing symbol after '\'")) 1))

beg f = (parse_string "\x=")
end f = (parse_string "\x=")
end g = (C ((L ((L I) "Missing definition")) 1))

beg f = (parse_string "\x=4")
end f = (parse_string "\x=4")
end g = ((R C) ((L I) ((S ((R R) ((L ((L ((L I) F)) "4")) 1))) (C ((L I) (C I))))))

beg f = (parse_string "\x\y\z x z; y z")
end f = (parse_string "\x\y\z x z; y z")
end g = ((R C) ((L I) (C ((L I) S))))

beg f = (parse_string "\x\y\z x; y z")
end f = (parse_string "\x\y\z x; y z")
end g = ((R C) ((L I) (C ((L I) R))))

beg f = (parse_string "\x\y\z x z y")
end f = (parse_string "\x\y\z x z y")
end g = ((R C) ((L I) (C ((L I) L))))

beg f = (parse_string "\x x")
end f = (parse_string "\x x")
end g = ((R C) ((L I) (C ((L I) I))))

beg f = (parse_string "\x\y x")
end f = (parse_string "\x\y x")
end g = ((R C) ((L I) (C ((L I) C))))

beg f = (parse_string "\x x x")
end f = (parse_string "\x x x")
end g = ((R C) ((L I) (C ((L I) ((S I) I)))))

beg f = (parse_string "\x=y x")
end f = (parse_string "\x=y x")
end g = ((R C) ((L I) ((S ((R R) ((L ((L ((L I) F)) "y")) 1))) (C ((L I) I)))))

beg f = (parse_string "\x=y y")
end f = (parse_string "\x=y y")
end g = ((R C) ((L I) ((S ((R R) ((L ((L ((L I) F)) "y")) 1))) (C ((L I) ((S C) I))))))

beg f = (parse_string "\y\x=y x x")
end f = (parse_string "\y\x=y x x")
end g = ((R C) ((L I) (C ((L I) ((S I) I)))))

beg f = (parse_string "\y\x=y y y")
end f = (parse_string "\y\x=y y y")
end g = ((R C) ((L I) (C ((L I) ((S ((R C) ((S I) I))) I)))))

beg f = (parse_string "()")
end f = (parse_string "()")
end g = ((R C) ((L I) (C ((L I) I))))

beg f = (parse_string "(
a
(b")
end f = (parse_string "(
a
(b")
end g = (C ((L ((L I) "Unclosed parenthesis")) 3))

beg f = (parse_string "(
a
(b)")
end f = (parse_string "(
a
(b)")
end g = (C ((L ((L I) "Unclosed parenthesis")) 1))

beg f = (parse_string "(
a
(b))")
end f = (parse_string "(
a
(b))")
end g = ((R C) ((L I) ((S ((R R) ((L ((L ((L I) F)) "b")) 3))) ((S ((R R) ((L ((L ((L I) F)) "a")) 2))) (C ((L I) I))))))

beg f = (parse_string "(ab cd e)
)")
end f = (parse_string "(ab cd e)
)")
end g = (C ((L ((L I) "Extraneous input")) 2))

beg f = (parse_file "missing.fxl")
end f = (parse_file "missing.fxl")
end g = (C ((L ((L I) "Missing file")) 1))

beg f = (parse_file "test/a2.fxl")
end f = (parse_file "test/a2.fxl")
end g = ((R C) ((L I) ((S ((R R) ((L ((L ((L I) C)) """)) 1))) ((S ((R R) ((L ((L ((L I) F)) "nl")) 7))) ((S ((R R) ((L ((L ((L I) F)) "put")) 6))) ((S ((R R) ((L ((L ((L I) C)) "ch = ")) 5))) ((S ((R R) ((L ((L ((L I) C)) "")) 4))) ((S ((R R) ((L ((L ((L I) F)) "eq")) 4))) ((S ((R R) ((L ((L ((L I) F)) "get")) 3))) ((S ((R R) ((L ((L ((L I) F)) "@")) 2))) (C ((L I) ((L ((R R) ((R R) ((R R) ((R R) ((R R) ((R _) _))))))) _)))))))))))))

beg f = (parse_file "test/utf8.fxl")
end f = (parse_file "test/utf8.fxl")
end g = ((R C) ((L I) ((S ((R R) ((L ((L ((L I) C)) "hej
åabcüdef
ghij
üä
1≠0
")) 11))) ((S ((R R) ((L ((L ((L I) C)) "hej åabcüdef")) 18))) ((S ((R R) ((L ((L ((L I) C)) "hej ghij")) 19))) ((S ((R R) ((L ((L ((L I) C)) "hej üä")) 20))) ((S ((R R) ((L ((L ((L I) C)) "hej 1≠0")) 21))) ((S ((R R) ((L ((L ((L I) C)) "grönsak")) 29))) ((S ((R R) ((L ((L ((L I) C)) "efterrätt")) 30))) ((S ((R R) ((L ((L ((L I) F)) "say")) 32))) ((S ((R R) ((L ((L ((L I) C)) "dessert ")) 32))) ((S ((R R) ((L ((_ _) _)) _))) _))))))))))))

beg f = (parse_string "say "hello"
say something
")
end f = (parse_string "say "hello"
say something
")
end g = ((R C) ((L I) ((S ((R R) ((L ((L ((L I) F)) "something")) 2))) ((S ((R R) ((L ((L ((L I) F)) "say")) 2))) ((S ((R R) ((L ((L ((L I) C)) "hello")) 1))) (C ((L I) ((L ((R S) (L I))) I))))))))

beg f = (is_str I)
end f = (is_str I)
end g = F

beg f = (is_str 4)
end f = (is_str 4)
end g = F

beg f = (is_str "x")
end f = (is_str "x")
end g = C

beg f = (is_str ((concat "x") "y"))
end f = (is_str "xy")
end g = C

beg f = (is_num "x")
end f = (is_num "x")
end g = F

beg f = (is_num 4.2)
end f = (is_num 4.2)
end g = C

beg f = (is_num ((* 3.9) 4.2))
end f = (is_num 16.38)
end g = C

beg f = (is_num (str_num "4.87"))
end f = (is_num 4.87)
end g = C

beg f = (is_num (str_num "  4.87"))
end f = (is_num 4.87)
end g = C

beg f = (is_num (str_num " 4.87 "))
end f = (is_num (str_num " 4.87 "))
end g = F

beg f = (is_num (str_num "4.87x"))
end f = (is_num (str_num "4.87x"))
end g = F

